plugins/core/hu.bme.mit.gamma.expression.model/src/hu/bme/mit/gamma/expression/util/ExpressionTypeDeterminator2.java

```diff
@@ -0,0 +1,361 @@
+		}
+		if (expression instanceof SubtractExpression) {
+			return getArithmeticBinaryType((SubtractExpression) expression);
+		}
+		if (expression instanceof DivideExpression) {
+			return getArithmeticBinaryType((DivideExpression) expression);
+		}
+		if (expression instanceof ModExpression) {
+			return getArithmeticBinaryIntegerType((ModExpression) expression);
+		}
+		if (expression instanceof DivExpression) {
+			return getArithmeticBinaryIntegerType((DivExpression) expression);
+		}
+		if (expression instanceof AddExpression) {
+			return getArithmeticMultiaryType((AddExpression) expression);
+		}
+		if (expression instanceof MultiplyExpression) {
+			return getArithmeticMultiaryType((MultiplyExpression) expression);
+		}
+		if (expression instanceof ArrayAccessExpression) {
+			Expression tmpExpression = ((ArrayAccessExpression) expression).getOperand();
+			Type declarationType = expressionUtil.getDeclaration(tmpExpression).getType();
+			if (declarationType instanceof ArrayTypeDefinition) {
+				return ((ArrayTypeDefinition) declarationType).getElementType();
+			}
+		}
+		if (expression instanceof FunctionAccessExpression) {
+			Type declarationType = expressionUtil.getDeclaration(expression).getType();
+			return ExpressionModelDerivedFeatures.getTypeDefinition(declarationType);
+		}
+		if (expression instanceof RecordAccessExpression) {
+			RecordAccessExpression recordAccessExpression = (RecordAccessExpression) expression;
+			return getType(recordAccessExpression.getFieldReference());
+		}
+		if (expression instanceof SelectExpression) {
+			Expression tmpExpression = ((SelectExpression) expression).getOperand();
+			Type declarationType = expressionUtil.getDeclaration(tmpExpression).getType();
+			TypeDefinition typeDefinition = removeTypeReferences(declarationType);
+			if (typeDefinition instanceof ArrayTypeDefinition) {
+				ArrayTypeDefinition arrayTypeDefinition = (ArrayTypeDefinition) typeDefinition;
+				return removeTypeReferences(arrayTypeDefinition.getElementType());
+			}
+			else if (typeDefinition instanceof IntegerRangeTypeDefinition) {
+				return factory.createIntegerTypeDefinition();
+			}
+			else if (typeDefinition instanceof EnumerationTypeDefinition) {
+				return typeDefinition;
+			}
+			else {
+				throw new IllegalArgumentException("The type of the operand of the select expression is not an enumerable type: " + expressionUtil.getDeclaration(expression));
+			}
+		}
+		if (expression == null) {
+			return factory.createVoidTypeDefinition();
+		}
+		throw new IllegalArgumentException("Unknown type!");
+	}
+	
+	// Equals, TypeReference
+	
+	public TypeDefinition removeTypeReferences(Type type) {
+		Type clone = ecoreUtil.clone(type);
+		for (TypeReference reference : ecoreUtil.getAllContentsOfType(clone, TypeReference.class)) {
+			TypeDefinition typeDefinition = ExpressionModelDerivedFeatures.getTypeDefinition(reference);
+			ecoreUtil.replace(reference, typeDefinition);
```

Comment:
Shouldn't the first object be the new object: `ecoreUtil.replace(typeDefinition, reference);`?